Fix Swift Doc-Link Generation

InsertCreativityHere opened the pull request:
This PR updates the `swiftLinkFormatter` to actually work.
To link a symbol in DocC, you wrap it in double backticks (not what we're doing right now).
There's also some small cleanup included in this PR as well.

I tested linking to data members, operations, enumerators, basically everything except for modules and parameters.
Note that linking to parameters is already disallowed by the parser.
I did stumble across a small bug with linking to constants, but the bug is with how we generate constants, not the link generation logic. See my blog post for more explanation.

----
----

Here's some pictures of the generated docs in a before/after format:
<img width="854" alt="Screenshot 2025-07-01 at 09 57 27" src="https://github.com/user-attachments/assets/205a07b2-34df-4ea9-9686-3249a6fca86b" />
<img width="850" alt="Screenshot 2025-07-01 at 09 57 32" src="https://github.com/user-attachments/assets/666ab253-b9d4-4ff5-83a0-7e216579b52d" />

----
----
<img width="840" alt="Screenshot 2025-07-01 at 09 56 35" src="https://github.com/user-attachments/assets/634f9b14-ea3e-46a4-a8dd-aab22fb46d93" />
<img width="842" alt="Screenshot 2025-07-01 at 09 56 45" src="https://github.com/user-attachments/assets/4f5758e4-436e-4fa1-bb20-f41c87b8005a" />

----
----
<img width="874" alt="Screenshot 2025-07-01 at 09 54 00" src="https://github.com/user-attachments/assets/8b9d887e-a00f-43b5-bcd7-117bd7372a24" />
<img width="833" alt="Screenshot 2025-07-01 at 09 54 52" src="https://github.com/user-attachments/assets/1815f2da-e74a-491a-8493-2c486d230f55" />

Note that when linking to interfaces, we always link to the proxy type.

externl commented:
I think we shouldn't generate warnings and should just not include a link if we don't know it.

bernardnormier commented:
> I think we shouldn't generate warnings and should just not include a link if we don't know it.

If you have doc-comment with `{@link Symbol}` and the parser can't resolve Symbol, it should definitely generate a warning.

Then, a secondary question is what happens next during code generation after such a warning: the code generator should not warn again. And it's ok for the code generator to not generate a proper link in this case.

pepone commented on cpp/src/slice2swift/SwiftUtil.cpp:
```diff
@@ -15,28 +15,68 @@ using namespace IceInternal;
+
+            // If the link was to an interface definition, we need to switch to the corresponding declaration.
+            // The code-gen considers `Def` the servant type, and `Decl` the proxy type. We want to link to the proxy.
+            if (auto interfaceDef = dynamic_pointer_cast<InterfaceDef>(contained))
```
is this block also used for operations, after they move up? If it is not an `else if` would make the intention clear, if it is the comment should make it clear.

InsertCreativityHere commented:
I rewrote the link logic.
Now we only emit a link if the source and target are in the same module.
Since DocC can't handle cross-module linking (in general).

Also updated some comments.

pepone reviewed (approved):
Looks good!

InsertCreativityHere commented on cpp/src/slice2swift/SwiftUtil.cpp:
```diff
@@ -8,35 +8,82 @@
+        }
+        else // We couldn't resolve the link target and make a best-effort attempt to map the raw link.
+        {
+            // For Swift all we can do is replace any doxygen separators with the DocC separator ('/').
```
Yep, 100%!
https://github.com/zeroc-ice/ice/blob/7fbfb6f2988bc1ede37d3c06c4b61c4061700ae7/cpp/src/Slice/Parser.cpp#L470-L484

InsertCreativityHere commented on cpp/src/slice2swift/SwiftUtil.cpp:
```diff
@@ -8,35 +8,82 @@
+            const string targetModule = Slice::Swift::getSwiftModule(contained->getTopLevelModule());
+
+            // Get the mapped-and-qualified name of the thing we're linking to, but replace any '.' with '/'.
+            // DocC uses forward slashes to separate symbols instead of periods.
```
Oh, that makes sense actually.
